### ISSUE TITLE
Referencing the implementing type in fixed interface choices

### ISSUE THREAD

**Issue by cocreature** [digital-asset/daml#11257]
Placeholder issue so I don’t forget. Will write down more details later today.

**Comment by sofiafaro-da** [comment]
> In Daml-LF, we need a way to represent that the type parameter is constrained by the interface.

I don't think we need this actually. You do need to know `selfType` at runtime to do virtual dispatch, and you can do that with parametrized typereps, which would be simpler IMO. This implies changing some primitives to accept parametrized typereps.

(nitpick/bikeshed: It would be cleaner to call it `thisType` instead of `selfType`. Currently, the type of `this` is `selfType` and the type of `self` is `ContractID selfType`. Versus, the type of `this` being `thisType` and `self` being `ContractID thisType`.)

**Comment by cocreature** [comment]
Sounds plausible. I’m a bit worried about introducing more and more crashes into daml-lf but I also don’t have a concrete issue I can point to that is caused by that.

**Comment by sofiafaro-da** [comment]
I'm not sure that any of these approaches is worth the added complexity. The argument that exposing `selfType` in the choice body would be more type-safe than just using the interface typereps from #11248 is a bit suspicious*. I would like better reason to pursue exposing `selfType` in the interface choice body.

(*- The argument requires that correctly converting a `Token` into a `selfType`, instead of using the `Token` methods directly, is easy, and it requires that comparing two interface typereps correctly is hard. This seems unlikely to me.)
